azavea/raster-foundry
# Remove Many-to-Many Relationship between Users and Organizations

## Overview

Removes `user_to_organizations` mapping table, and moves its fields to the `users` table. Also updates corresponding data models, Slick table expressions, routes, tests, and spec. See commit messages for details.

### Checklist

~~- [ ] Styleguide updated, if necessary~~

### Demo

###### New `users` table:

```sql
SELECT * FROM users;

                 id                  |           organization_id            |  role  |         created_at         |        modified_at
-------------------------------------+--------------------------------------+--------+----------------------------+----------------------------
 default                             | dfac6307-b5ef-43f7-beda-b9f208bb7726 | VIEWER | 2017-03-16 19:17:27.252158 | 2017-03-16 19:17:27.252158
 rf|airflow-user                     | dfac6307-b5ef-43f7-beda-b9f208bb7726 | VIEWER | 2017-03-16 19:17:27.252158 | 2017-03-16 19:17:27.252158
 auth0|xxxxxxxxxxxxxxxxxxxxxxxx      | dfac6307-b5ef-43f7-beda-b9f208bb7726 | VIEWER | 2017-03-16 19:17:27.252158 | 2017-03-16 19:17:27.252158
 google-oauth2|111111111111111111111 | dfac6307-b5ef-43f7-beda-b9f208bb7726 | VIEWER | 2017-03-16 19:17:27.252158 | 2017-03-16 19:17:27.252158
 google-oauth2|222222222222222222222 | dfac6307-b5ef-43f7-beda-b9f208bb7726 | VIEWER | 2017-03-16 19:17:27.252158 | 2017-03-16 19:17:27.252158
```

###### Getting `PaginatedResponse[User.WithOrg]` from `/users/` and `User.WithOrg` from `/users/{id}`

```http
$ http --print HhBb :9100/api/users Authorization:"Bearer $RF_JWT"

GET /api/users HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Authorization: Bearer eyJ...
Connection: keep-alive
Host: localhost:9100
User-Agent: HTTPie/0.9.8



HTTP/1.1 200 OK
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Fri, 17 Mar 2017 20:22:01 GMT
Server: nginx
Strict-Transport-Security: max-age=15552000; preload
Transfer-Encoding: chunked
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block

{
    "count": 24,
    "hasNext": false,
    "hasPrevious": false,
    "page": 0,
    "pageSize": 30,
    "results": [
        {
            "createdAt": "2017-03-16T19:17:27.252158Z",
            "id": "default",
            "modifiedAt": "2017-03-16T19:17:27.252158Z",
            "organization": {
                "createdAt": "2016-11-16T01:55:32.916196Z",
                "id": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
                "modifiedAt": "2016-11-16T01:55:32.916196Z",
                "name": "Public"
            },
            "role": "Viewer"
        },
        {
            "createdAt": "2017-03-16T19:17:27.252158Z",
            "id": "rf|airflow-user",
            "modifiedAt": "2017-03-16T19:17:27.252158Z",
            "organization": {
                "createdAt": "2016-11-16T01:55:32.916196Z",
                "id": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
                "modifiedAt": "2016-11-16T01:55:32.916196Z",
                "name": "Public"
            },
            "role": "Viewer"
        },
        {
            "createdAt": "2017-03-16T19:17:27.252158Z",
            "id": "auth0|xxxxxxxxxxxxxxxxxxxxxxxx",
            "modifiedAt": "2017-03-16T19:17:27.252158Z",
            "organization": {
                "createdAt": "2016-11-16T01:55:32.916196Z",
                "id": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
                "modifiedAt": "2016-11-16T01:55:32.916196Z",
                "name": "Public"
            },
            "role": "Viewer"
        }
    ]
}
```

```http
$ http --print HhBb ":9100/api/users/google-oauth2|222222222222222222222" Authorization:"Bearer $RF_JWT"

GET /api/users/google-oauth2%7C222222222222222222222 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Authorization: Bearer eyJ...
Connection: keep-alive
Host: localhost:9100
User-Agent: HTTPie/0.9.8



HTTP/1.1 200 OK
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Fri, 17 Mar 2017 20:24:54 GMT
Server: nginx
Strict-Transport-Security: max-age=15552000; preload
Transfer-Encoding: chunked
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block

{
    "createdAt": "2017-03-16T19:17:27.252158Z",
    "id": "google-oauth2|222222222222222222222",
    "modifiedAt": "2017-03-16T19:17:27.252158Z",
    "organization": {
        "createdAt": "2016-11-16T01:55:32.916196Z",
        "id": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
        "modifiedAt": "2016-11-16T01:55:32.916196Z",
        "name": "Public"
    },
    "role": "Viewer"
}
```

###### Updating a User with `User.Update`

```http
$ http --print HhBb PUT ":9100/api/users/google-oauth2|222222222222222222222" Authorization:"Bearer $RF_JWT" organizationId=dfac6307-b5ef-43f7-beda-b9f208bb7726 role=User

PUT /api/users/google-oauth2%7C222222222222222222222 HTTP/1.1
Accept: application/json, */*
Accept-Encoding: gzip, deflate
Authorization: Bearer eyJ...
Connection: keep-alive
Content-Length: 74
Content-Type: application/json
Host: localhost:9100
User-Agent: HTTPie/0.9.8

{
    "organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
    "role": "User"
}

HTTP/1.1 204 No Content
Connection: keep-alive
Date: Fri, 17 Mar 2017 20:26:56 GMT
Server: nginx
Strict-Transport-Security: max-age=15552000; preload
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block


```

### Notes

Should there be new tests for this?

I created `User.WithOrg` for convenience's sake, but am not sure if we should be using it everywhere. If so desired, we can easily switch to `User`. The only difference between the two is that the former includes the entire Organization, as opposed to only the Organization Id in the latter.

## Testing Instructions

 * Check out this branch
 * Ensure you can list users and change their roles
 * Ensure that the rest of the app works as expected

Closes #1252
